diff --git a/addons/account/account_move_line.py b/addons/account/account_move_line.py
index 84778dc..494e178 100644
--- a/addons/account/account_move_line.py
+++ b/addons/account/account_move_line.py
@@ -19,6 +19,7 @@
 #
 ##############################################################################
 
+import logging
 import time
 from datetime import datetime
 
@@ -30,6 +31,8 @@ from openerp import tools
 from openerp.report import report_sxw
 import openerp
 
+_logger = logging.getLogger(__name__)
+
 class account_move_line(osv.osv):
     _name = "account.move.line"
     _description = "Journal Items"
@@ -167,6 +170,7 @@ class account_move_line(osv.osv):
             result = move_line_total
             res[move_line.id]['amount_residual_currency'] =  sign * (move_line.currency_id and self.pool.get('res.currency').round(cr, uid, move_line.currency_id, result) or result)
             res[move_line.id]['amount_residual'] = sign * line_total_in_company_currency
+            _logger.info("_amount_residual %s: amount_residual_currency %s: move_line_total %s: sign %s:" % (res[move_line.id]['amount_residual'], res[move_line.id]['amount_residual_currency'], move_line_total, sign))
         return res
 
     def default_get(self, cr, uid, fields, context=None):
@@ -512,7 +516,8 @@ class account_move_line(osv.osv):
         'account_tax_id':fields.many2one('account.tax', 'Tax', copy=False),
         'analytic_account_id': fields.many2one('account.analytic.account', 'Analytic Account'),
         'company_id': fields.related('account_id', 'company_id', type='many2one', relation='res.company',
-                            string='Company', store=True, readonly=True)
+                            string='Company', store=True, readonly=True),
+        'account_separate': fields.boolean(string='Separate movement', default=False)
     }
 
     def _get_date(self, cr, uid, context=None):